### DISCUSSION TITLE
Verify email address before persisting it

### DISCUSSION THREAD

**Discussion by alantbarlow** [ory/kratos#2655] [Q&A] [answered]
I want to require email verification during the registration process. The flow I want to do is the following, 

1. The user submits an email address for verification.
2. The user verifies the email address
3. The user finishes registration by selecting login method (Webauthn)

Im wanting these steps to occur before the user is registered. It sounds like the verification can only happen on "known" addresses which defeats the purpose of email verification. I need to be able to prevent sign-up with an un-verified email address.

Im also wanting to do something similar for the user changing their email address. From what I can see is that also is not possible for the same reason. I want the email to be verified before persisting the address into the database. That means before registration or before updating the registered email address.

Can you please let me know if this is possible?

**Comment by vinckr** [comment]
Hello @alantbarlow 
You can prevent the user from logging in without a verified email address.
So only accounts whose email has been verified can log in after signing up.
AFAIK there is no way to prevent them from creating an account - but with the above that would have the same results from the user perspective from my POV. You can then purge all accounts without verified email.

>verification can only happen on "known" addresses which defeats the purpose of email verification

Can you elaborate a bit on that? What do you mean by known addresses? The mail courier has to know the email to deliver the verification mail. 

There are also [webhooks](https://www.ory.sh/docs/kratos/self-service/hooks) that run before persisting -  you can build all kinds of custom logic with that if the above is not a good solution for you. Finally if there is no good practical solution possible right now, we are always open to feature requests and contributions!

**Reply by alantbarlow** [reply]
> You can prevent the user from logging in without a verified email address.
So only accounts whose email has been verified can log in after signing up.

I knew this was possible but it won't prevent a signup from someone using someone else's email address. With the assumption that a webauthn device can not be changed through account recovery, this would prevent the user from being able to create an account or recover the account that was created for them.

>  You can then purge all accounts without verified email.

This could solve the above issue but require a scheduled process to keep the database clean from fake accounts. This obviously takes additional server resources to accomplish and would also purge accounts where the user changed their email after signing up but never verified the new email address.

> Can you elaborate a bit on that? What do you mean by known addresses? The mail courier has to know the email to deliver the verification mail.

<img width="980" alt="image" src="https://user-images.githubusercontent.com/15796492/183899901-c50ac8dc-b59b-471b-b50c-bc8f1817f53b.png">

This is what I mean by known or unknown address. If the email verification API provided a way to verify email addresses before persisting them, then we could prevent fake sign-ups which would save server resources from being used by the need to "purge the system" of the fake accounts. We would prevent real accounts from being "purged" that haven't verified their email because every email would be pre-verified. This would also prevent users from getting locked out of their account by mistyping their email address when updating it.

The documentation also suggests that the need for a backup email address would be needed to prevent account lockout when the email address is changed. This seems to be the only workaround without being able to pre-verify the email before persisting the change.

With all that said, the suggested workarounds of periodically purging the system of unverified email addresses, and adding/removing email addresses instead of changing them could solve these issues. 

If you require email verification, Is the user able to sign-in with any verified email address on their account, or does it have to be the primary email address? If it has to be the primary, is there an easy way for the user to change which email address is the primary one and do it automatically if the primary email address is deleted?

**Reply by Benehiko** [reply]
Hi @alantbarlow 

Yes you are correct, you cannot recover an account that has webauthn linked to it.
This is a bug and we should convert this to an issue then.

I could also reproduce locking myself out of my account by doing the following flow:

1. Create account with 2FA
2. Add Yubikey
3. Add backup codes
4. Remove Yubikey, but keep backup codes
5. Logout
6. Try recover account
7. Stuck at backup codes screen

**Reply by Benehiko** [reply, marked as answer]
Keeping track of this https://github.com/ory/kratos/issues/2656
